metrics-core/src/main/java/com/codahale/metrics/LockFreeExponentiallyDecayingReservoir.java

```diff
@@ -0,0 +1,246 @@
+                newCount = new AtomicLong(consumer.count);
+            } else {
+                newCount = new AtomicLong();
+            }
+            return new State(newTick, newCount, newValues);
+        }
+    }
+
+    private static final class RescalingConsumer implements BiConsumer<Double, WeightedSample> {
+        private final double scalingFactor;
+        private final ConcurrentSkipListMap<Double, WeightedSample> values;
+        private long count;
+
+        RescalingConsumer(double scalingFactor, ConcurrentSkipListMap<Double, WeightedSample> values) {
+            this.scalingFactor = scalingFactor;
+            this.values = values;
+        }
+
+        @Override
+        public void accept(Double priority, WeightedSample sample) {
+            double newWeight = sample.weight * scalingFactor;
+            if (Double.compare(newWeight, 0) == 0) {
+                return;
+            }
+            WeightedSample newSample = new WeightedSample(sample.value, newWeight);
+            if (values.put(priority * scalingFactor, newSample) == null) {
+                count++;
+            }
+        }
+    }
+
+    private LockFreeExponentiallyDecayingReservoir(int size, double alpha, Duration rescaleThreshold, Clock clock) {
+        this.alpha = alpha;
+        this.size = size;
+        this.clock = clock;
+        this.rescaleThresholdNanos = rescaleThreshold.toNanos();
+        this.state = new State(clock.getTick(), new AtomicLong(), new ConcurrentSkipListMap<>());
+    }
+
+    @Override
+    public int size() {
+        return (int) Math.min(size, state.count.get());
+    }
+
+    @Override
+    public void update(long value) {
+        long now = clock.getTick();
+        rescaleIfNeeded(now).update(value, now);
+    }
+
+    private State rescaleIfNeeded(long currentTick) {
+        State stateSnapshot = this.state;
+        long lastScaleTick = stateSnapshot.startTick;
+        if (currentTick - lastScaleTick >= rescaleThresholdNanos) {
+            State newState = stateSnapshot.rescale(currentTick);
+            if (stateUpdater.compareAndSet(this, stateSnapshot, newState)) {
+                // newState successfully installed
+                return newState;
+            }
+            // Otherwise another thread has won the race and we can return the result of a volatile read.
+            // It's possible this has taken so long that another update is required, however that's unlikely
+            // and no worse than the standard race between a rescale and update.
+            return this.state;
+        }
+        return stateSnapshot;
+    }
+
+    @Override
+    public Snapshot getSnapshot() {
+        State stateSnapshot = rescaleIfNeeded(clock.getTick());
+        return new WeightedSnapshot(stateSnapshot.values.values());
+    }
+
+    private double weight(long durationNanos) {
+        double durationSeconds = durationNanos / 1_000_000_000D;
+        return Math.exp(alpha * durationSeconds);
+    }
```

Comment:
Note that this differs from the original ExponentiallyDecayingReservoir by avoiding a cliff every second. Nanosecond values are converted to seconds as a double for smooth scaling.